Mathlib/Algebra/Ring/Defs.lean

```diff
@@ -0,0 +1,471 @@
+-- see Note [lower instance priority]
+instance (priority := 100) Distrib.rightDistribClass (R : Type _) [Distrib R] :
+    RightDistribClass R :=
+  ⟨Distrib.right_distrib⟩
+#align distrib.right_distrib_class Distrib.rightDistribClass
+
+theorem left_distrib [Mul R] [Add R] [LeftDistribClass R] (a b c : R) :
+    a * (b + c) = a * b + a * c :=
+  LeftDistribClass.left_distrib a b c
+#align left_distrib left_distrib
+
+alias left_distrib ← mul_add
+
+theorem right_distrib [Mul R] [Add R] [RightDistribClass R] (a b c : R) :
+    (a + b) * c = a * c + b * c :=
+  RightDistribClass.right_distrib a b c
+#align right_distrib right_distrib
+
+alias right_distrib ← add_mul
+
+theorem distrib_three_right [Mul R] [Add R] [RightDistribClass R] (a b c d : R) :
+    (a + b + c) * d = a * d + b * d + c * d := by simp [right_distrib]
+#align distrib_three_right distrib_three_right
+
+/-!
+### Semirings
+-/
+
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring. -/
+class NonUnitalNonAssocSemiring (α : Type u) extends AddCommMonoid α, Distrib α, MulZeroClass α
+#align non_unital_non_assoc_semiring NonUnitalNonAssocSemiring
+
+/-- An associative but not-necessarily unital semiring. -/
+class NonUnitalSemiring (α : Type u) extends NonUnitalNonAssocSemiring α, SemigroupWithZero α
+#align non_unital_semiring NonUnitalSemiring
+
+/-- A unital but not-necessarily-associative semiring. -/
+class NonAssocSemiring (α : Type u) extends NonUnitalNonAssocSemiring α, MulZeroOneClass α,
+    AddCommMonoidWithOne α
+#align non_assoc_semiring NonAssocSemiring
+
+class Semiring (α : Type u) extends NonUnitalSemiring α, NonAssocSemiring α, MonoidWithZero α
+#align semiring Semiring
+
+section AddMonoidWithOne
+
+variable [AddMonoidWithOne α]
+
+-- TODO: Move to `Data.Nat.Cast.Defs`?
+theorem one_add_one_eq_two : 1 + 1 = (2 : α) := by
+  rw [←Nat.cast_one, ←Nat.cast_add]
+  apply congrArg
+  decide
+#align one_add_one_eq_two one_add_one_eq_two
+
+end AddMonoidWithOne
+
+section DistribMulOneClass
+
+variable [Add α] [MulOneClass α]
+
+theorem add_one_mul [RightDistribClass α] (a b : α) : (a + 1) * b = a * b + b := by
+  rw [add_mul, one_mul]
+#align add_one_mul add_one_mul
+
+theorem mul_add_one [LeftDistribClass α] (a b : α) : a * (b + 1) = a * b + a := by
+  rw [mul_add, mul_one]
+#align mul_add_one mul_add_one
+
+theorem one_add_mul [RightDistribClass α] (a b : α) : (1 + a) * b = b + a * b := by
+  rw [add_mul, one_mul]
+#align one_add_mul one_add_mul
+
+theorem mul_one_add [LeftDistribClass α] (a b : α) : a * (1 + b) = a + a * b := by
+  rw [mul_add, mul_one]
+#align mul_one_add mul_one_add
+
+end DistribMulOneClass
+
+section Semiring
+
+variable [Semiring α]
+
+-- Porting note: was [has_add α] [mul_one_class α] [right_distrib_class α]
+theorem two_mul (n : α) : 2 * n = n + n :=
+  (congrArg₂ _ one_add_one_eq_two.symm rfl).trans <| (right_distrib 1 1 n).trans (by rw [one_mul])
+#align two_mul two_mul
+
+section deprecated
+set_option linter.deprecated false
```

Comment:
If it's only for one command, you can use

`set_option linter.deprecated false in ...`